File: frontend/src/types/budget.ts
// ---------- Budget Category ----------

export interface BudgetCategory {
  id: number;
  name: string;
  slug: string;
  display_order: number;
  is_system: boolean;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface BudgetCategoryPayload {
  name: string;
  slug: string;
  display_order: number;
  is_active?: boolean;
}

export interface BudgetCategoryUpdatePayload {
  name?: string;
  slug?: string;
  display_order?: number;
  is_active?: boolean;
}

// ---------- Budget Item ----------

export interface CategorySummary {
  id: number;
  name: string;
  slug: string;
}

export interface BudgetItem {
  id: number;
  project_id: number;
  category_id: number;
  category: CategorySummary;
  description: string;
  planned_amount: string;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface BudgetItemPayload {
  category_id: number;
  description: string;
  planned_amount: number;
  notes?: string | null;
}

export interface BudgetItemUpdatePayload {
  category_id?: number;
  description?: string;
  planned_amount?: number;
  notes?: string | null;
}

// ---------- Budget Summary ----------

export interface BudgetCategoryBreakdown {
  category_id: number;
  category_name: string;
  category_slug: string;
  planned_amount: string;
  spent_amount: string;
  remaining_amount: string;
  utilization_pct: number;
}

export interface BudgetSummary {
  project_id: number;
  project_budget_rub: string;
  total_planned: string;
  total_spent: string;
  total_pending: string;
  remaining: string;
  utilization_pct: number;
  by_category: BudgetCategoryBreakdown[];
}

// ---------- Expense ----------

export interface Expense {
  id: number;
  project_id: number;
  budget_item_id: number | null;
  category_id: number;
  category: CategorySummary;
  description: string;
  amount: string;
  expense_date: string;
  vendor: string | null;
  invoice_number: string | null;
  notes: string | null;
  status: string;
  creator_name: string;
  created_at: string;
  updated_at: string;
}

export interface ExpensePayload {
  category_id: number;
  budget_item_id?: number | null;
  description: string;
  amount: number;
  expense_date: string;
  vendor?: string | null;
  invoice_number?: string | null;
  notes?: string | null;
}

export interface ExpenseUpdatePayload {
  category_id?: number;
  budget_item_id?: number | null;
  description?: string;
  amount?: number;
  expense_date?: string;
  vendor?: string | null;
  invoice_number?: string | null;
  notes?: string | null;
}

export interface ImportRowError {
  row: number;
  reason: string;
}

export interface ExpenseImportResult {
  imported_count: number;
  skipped_count: number;
  errors: ImportRowError[];
}

File: backend/app/schemas/budget.py
"""Pydantic schemas for BudgetItem domain.

This module declares the request/response shapes used by the budget items
API. Variance schemas (planned vs actual) live in
``app.schemas.budget_variance`` so this file stays compact -- the dev
sandbox occasionally truncates larger files mid-write, so we keep the
hot-path schema module short and pad the tail with comments to ensure
we always overwrite any stale content from a previous version.

Naming conventions
------------------
* ``*Base``      -- shared fields between create + update
* ``*Create``    -- create payload (POST body)
* ``*Update``    -- partial update payload (PATCH/PUT body)
* ``*Response``  -- enriched read model (includes db-side fields)

Backwards-compat policy
-----------------------
New optional fields land here first. Existing endpoints continue to
return ``None`` until the migration that introduces the column has been
applied. See ``alembic/versions/c2d4e6f8a1b3_add_cost_code...`` for the
columns introduced by Faz 2.
"""
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, model_validator


class CategorySummary(BaseModel):
    """Minimal category info embedded in budget item responses."""

    id: int
    name: str
    slug: str
    model_config = ConfigDict(from_attributes=True)


class BudgetItemBase(BaseModel):
    """Shared fields between create and update payloads."""

    category_id: int | None = None
    category_name_new: str | None = Field(None, max_length=100)
    description: str = Field(..., min_length=1, max_length=500)
    cost_code: str | None = Field(None, max_length=50)
    planned_amount: Decimal = Field(default=Decimal("0"), ge=0)
    committed_amount: Decimal = Field(default=Decimal("0"), ge=0)
    notes: str | None = None

    @model_validator(mode="after")
    def _check_exactly_one_category(self):
        if (self.category_id is None) == (self.category_name_new is None):
            raise ValueError(
                "Provide exactly one of category_id or category_name_new"
            )
        return self


class BudgetItemCreate(BudgetItemBase):
    """Create payload for ``POST /projects/{id}/budget-items``."""

    pass


class BudgetItemUpdate(BaseModel):
    """Partial update payload. All fields optional."""

    category_id: int | None = None
    category_name_new: str | None = Field(None, max_length=100)
    description: str | None = Field(None, min_length=1, max_length=500)
    cost_code: str | None = Field(None, max_length=50)
    planned_amount: Decimal | None = Field(None, ge=0)
    committed_amount: Decimal | None = Field(None, ge=0)
    notes: str | None = None

    @model_validator(mode="after")
    def _check_at_most_one_category(self):
        if self.category_id is not None and self.category_name_new is not None:
            raise ValueError(
                "Cannot provide both category_id and category_name_new"
            )
        return self


class BudgetItemResponse(BudgetItemBase):
    """Read model: budget item + computed/db-side fields."""

    id: int
    project_id: int
    category: CategorySummary
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)


class BudgetCategoryBreakdown(BaseModel):
    """Aggregation per category for one project (used by BudgetSummary)."""

    category_id: int
    category_name: str
    category_slug: str
    planned_amount: Decimal
    spent_amount: Decimal
    remaining_amount: Decimal
    utilization_pct: float


class BudgetSummary(BaseModel):
    """Full budget summary returned by the per-project summary endpoint."""

    project_id: int
    project_budget_rub: Decimal
    total_planned: Decimal
    total_spent: Decimal
    total_pending: Decimal
    remaining: Decimal
    utilization_pct: float
    by_category: list[BudgetCategoryBreakdown]
    # Number of underlying ledger expense entries that contributed to
    # total_spent. Lets the UI render "N expense records" without
    # re-fetching the ledger list. Defaults to 0 for backwards-compat.
    expense_records_count: int = 0


class BudgetImportRowError(BaseModel):
    """A single row that failed validation during budget item import."""

    row: int
    reason: str


class BudgetImportRowWarning(BaseModel):
    """A row that was skipped or modified with a warning during import."""

    row: int
    reason: str


class BudgetImportResult(BaseModel):
    """Summary returned after a budget item Excel import."""

    imported_count: int
    skipped_count: int
    deleted_count: int = 0
    errors: list[BudgetImportRowError]
    warnings: list[BudgetImportRowWarning] = []


# -------------------------------------------------------------------------
# Padding to keep the file longer than any previously-truncated stale copy
# on disk. The sandbox we develop in occasionally fails to truncate the
# old content when a smaller new payload is written, leaving null-byte
# padding that breaks the Python parser. By keeping the file comfortably
# larger than its prior versions we avoid that failure mode.
# -------------------------------------------------------------------------
